spark-rapids/mig.sh

```diff
@@ -92,74 +265,168 @@ EOF
+      mkdir -p "${WORKDIR}"
+      pushd $_
+      # Fetch open souce kernel module with corresponding tag
+      git clone https://github.com/NVIDIA/open-gpu-kernel-modules.git \
+          --branch "${NVIDIA_DRIVER_VERSION}" --single-branch
+      cd ${WORKDIR}/open-gpu-kernel-modules
+      #
+      # build kernel modules
+      #
+      make -j$(nproc) modules \
+	   > /var/log/open-gpu-kernel-modules-build.log \
+	  2> /var/log/open-gpu-kernel-modules-build_error.log
+      configure_dkms_certs
+      # sign
+      for module in $(find kernel-open -name '*.ko'); do
+        /lib/modules/$(uname -r)/build/scripts/sign-file sha256 \
+          "${CA_TMPDIR}/db.rsa" \
+	  "${CA_TMPDIR}/db.der" \
+	  "${module}"
+      done
+      clear_dkms_key
+      # install
+      make modules_install \
+	   >> /var/log/open-gpu-kernel-modules-build.log \
+	  2>> /var/log/open-gpu-kernel-modules-build_error.log
+      depmod -a
+      modprobe nvidia
+      popd
+
+      #
+      # Install CUDA
+      #
+      cuda_runfile="cuda_${CUDA_VERSION}_${NVIDIA_DRIVER_VERSION}_linux.run"
+      curl -fsSL --retry-connrefused --retry 10 --retry-max-time 30 \
+       "https://developer.download.nvidia.com/compute/cuda/${CUDA_VERSION}/local_installers/${cuda_runfile}" \
+       -o cuda.run
+      bash cuda.run --silent --toolkit --no-opengl-libs
+      rm cuda.run
+    else
+      # Install from repo provided by NV
+      readonly UBUNTU_REPO_CUDA_PIN="${NVIDIA_REPO_URL}/cuda-${shortname}.pin"
+
+      curl -fsSL --retry-connrefused --retry 3 --retry-max-time 5 \
+        "${UBUNTU_REPO_CUDA_PIN}" -o /etc/apt/preferences.d/cuda-repository-pin-600
+
+      curl -fsSL --retry-connrefused --retry 3 --retry-max-time 5 \
+        "${LOCAL_DEB_URL}" -o /tmp/local-installer.deb
+
+      dpkg -i /tmp/local-installer.deb
+      rm /tmp/local-installer.deb
+      cp ${DIST_KEYRING_DIR}/cuda-*-keyring.gpg /usr/share/keyrings/
+      execute_with_retries "apt-get update"
+
+      execute_with_retries "apt-get install -y -q --no-install-recommends dkms"
+      configure_dkms_certs
+      for pkg in "nvidia-driver-${NVIDIA_DRIVER_VERSION_PREFIX}-open" \
+                 "cuda-drivers-${NVIDIA_DRIVER_VERSION_PREFIX}" \
+                 "cuda-toolkit-${CUDA_VERSION_MAJOR//./-}" ; do
+        execute_with_retries "apt-get install -y -q --no-install-recommends ${pkg}"
+      done
+      clear_dkms_key
+
+      modprobe nvidia
+    fi
 
-    dpkg -i /tmp/local-installer.deb
-    cp /var/cuda-repo-ubuntu${UBUNTU_VERSION}04-${CUDA_VERSION_MAJOR//./-}-local/cuda-*-keyring.gpg /usr/share/keyrings/
-    execute_with_retries "apt-get update"    
-    
-    execute_with_retries "apt-get install -y -q --no-install-recommends cuda-drivers-${NVIDIA_DRIVER_VERSION_PREFIX}"
-    execute_with_retries "apt-get install -y -q --no-install-recommends cuda-toolkit-${CUDA_VERSION_MAJOR//./-}"
 
     # enable a systemd service that updates kernel headers after reboot
```

Comment:
No, bad idea.  The kernel should not change.  To get a new kernel, upgrade the image.  The kernel version should be pinned, in my opinion.  Otherwise, there are too many failure states.

Reply:
I agree. Currently, reboots can trigger kernel upgrades, potentially breaking the drivers. Ideally, we should avoid kernel upgrades.

Reply:
I don't know if I can support customers running outdated images.  I guess we've been doing it to date by applying main force.  But it's not really sustainable.  Can you tell me whether your users will be able to move to a new dataproc image to get new kernels?

The customer can only apply new kernel fixes if they reboot into a new kernel.  When they do that, they will lose the drivers that I build for them just for that kernel version.  I do not enable DKMS because we do not recommend changing kernels.  Since DKMS is not installed, the system will not even attempt to build a new version of the kernel driver.  The assumption is that the service account is only granted access to the driver signing private key material during the installation, or more realistically, during custom image generation.

On systems with secure boot enabled, the user will not be able to fetch the signing material required for driver re-build.  It's just a lot less hassle if customers decommission old images and move to new images regularly.